File: scripts/optimize_images.py
import os
from pathlib import Path

from PIL import Image

# Configuration
INPUT_DIR = "public/images/board"
OUTPUT_DIR = "public/optimized-images/board"
TARGET_SIZE = (600, 600)  # This is larger than needed (300x300) to account for high-DPI displays


def optimize_image(input_path, output_path):
    # Create output directory if it doesn't exist
    os.makedirs(os.path.dirname(output_path), exist_ok=True)

    with Image.open(input_path) as img:
        # Keep transparency for RGBA images
        # Calculate dimensions maintaining aspect ratio
        width, height = img.size

        # Calculate new dimensions where smaller side is TARGET_SIZE
        if width < height:
            new_width = TARGET_SIZE[0]
            new_height = int(height * (TARGET_SIZE[0] / width))
        else:
            new_height = TARGET_SIZE[1]
            new_width = int(width * (TARGET_SIZE[1] / height))

        # Resize image
        img = img.resize((new_width, new_height), Image.Resampling.LANCZOS)

        # Save as WebP, preserving transparency if present
        img.save(output_path, "WEBP", quality=85)


def main():
    input_dir = Path(INPUT_DIR)
    output_dir = Path(OUTPUT_DIR)

    # Create output directory if it doesn't exist
    output_dir.mkdir(parents=True, exist_ok=True)

    # Process all images in input directory
    for input_path in input_dir.glob("*"):
        if input_path.suffix.lower() in (".png", ".jpg", ".jpeg", ".gif", ".webp"):
            output_path = output_dir / f"{input_path.stem}.webp"
            print(f"Processing: {input_path.name}")
            optimize_image(str(input_path), str(output_path))
            print(f"Saved to: {output_path.name}")


if __name__ == "__main__":
    main()

File: scripts/optimize-images.mjs
import { access, mkdir, readdir } from 'fs/promises';
import { join, parse } from 'path';
import sharp from 'sharp';

const INPUT_DIR = 'public/images/board';
const OUTPUT_DIR = 'public/optimized-images/board';
const TARGET_SIZE = 600;

async function fileExists (path) {
  try {
    await access(path);
    return true;
  } catch {
    return false;
  }
}

async function optimizeImage (inputPath, outputPath) {
  try {
    if (await fileExists(outputPath)) {
      return;
    }

    const image = sharp(inputPath);
    const metadata = await image.metadata();

    const { width, height } = metadata;
    const newDimensions = width < height
      ? { width: TARGET_SIZE, height: Math.round(height * (TARGET_SIZE / width)) }
      : { width: Math.round(width * (TARGET_SIZE / height)), height: TARGET_SIZE };

    await image
      .resize(newDimensions)
      .webp({ quality: 85 })
      .toFile(outputPath);

    console.log(`✓ Saved: ${outputPath}`);
  } catch (error) {
    console.error(`❌ Error processing ${inputPath}:`, error);
  }
}

async function main () {
  try {
    await mkdir(OUTPUT_DIR, { recursive: true });

    const files = await readdir(INPUT_DIR);
    const imageFiles = files.filter(file =>
      /\.(jpg|jpeg|png|webp|gif)$/i.test(file)
    );

    for (const file of imageFiles) {
      const inputPath = join(INPUT_DIR, file);
      const { name } = parse(file);
      const outputPath = join(OUTPUT_DIR, `${name}.webp`);

      await optimizeImage(inputPath, outputPath);
    }
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

main();